### PR TITLE
a bit of performance improvements

### PR DESCRIPTION
**Pull request by mamico**
1. removed marking request for disable CSRF protection (scalesdict are already "safe type" for plone.protect)
2. use the attribute _sizes for local caching correctly available_sizes
3. add additional infos in scale storage only if missing (probably could be better to move writes in plone.scale)

Tested against a @querystring-search restapi that returns fullobjects with leadimage:
before 0.12289s, with these changes 0.06316s (b_size=10) and from 2.02214s to 0.97613s (b_size=200)

### CODE REVIEW

**Comment by mamico** [comment]
@jenkins-plone-org please run jobs

File: plone/namedfile/scaling.py
Comment (mauritsvanrees) [inline plone/namedfile/scaling.py:421]: This condition is always true.

File: plone/namedfile/scaling.py
Comment (mauritsvanrees) [inline plone/namedfile/scaling.py:510]: So the `srcset` is actually stored in the `ScalesDict`? I did not expect that. You are right though.

A downside of not updating it always, is that the srcset configuration may have changed. Maybe 3x was not enabled earlier and now it is. Or the image quality at 2x was increased. You do not see the improvements here then.
Could be a minor issue.

Is this the part that gives the most performance improvement? I expect that this is mostly because the scale gets generated. Or did you try and measure after all the scales, including 2x and 3x were already generated? 

It may be good to wait until my improvements from PR #113 are approved (maybe by you) and merged. Then measure again. There I make a split between creating a unique id for the scale and actually generating the scale.

File: plone/namedfile/scaling.py
Comment (mamico) [reply plone/namedfile/scaling.py]: +1

File: plone/namedfile/scaling.py
Comment (mamico) [reply plone/namedfile/scaling.py:509]: > Is this the part that gives the most performance improvement? I expect that this is mostly because the scale gets generated. Or did you try and measure after all the scales, including 2x and 3x were already generated?
>
> It may be good to wait until my improvements from PR #113 are approved (maybe by you) and merged. Then measure again. There I make a split between creating a unique id for the scale and actually generating the scale.

Our tests are all within all the scale generated. So probably a different path from your improvements.

I finally agree with you, we can wait until you have merged your current works, happy to review them, and then reevaluate this.

**Comment by mamico** [comment]
@jenkins-plone-org please run jobs

**Comment by mamico** [comment]
@mauritsvanrees @cekk 
I saw that #113 was merged. So I synced the master code in the branch of this PR and reevaluated the performance improvements.
Same, good results as above. Surprisingly, most performance improvements come from the remove of `IDisableCSRFProtection`. 
But by removing that there is a need to modify the storage adapter in `plone.scale`, just for the purpose of testing this PR, I inserted a temporary monkey patch for this (https://github.com/plone/plone.namedfile/pull/117/commits/f71706a2f9b405b4902b658a75837c405abd5401). Eventually, if it's ok in the whole, I will move the monkey patch to a related PR in `plone.scale`.

**Comment by ale-rt** [comment]
Just my 2 c.

I think ScalesDict is not performant because it inherits from PersistentDict:
- https://github.com/plone/plone.scale/blob/258731910a35f37056a2097d7607e35db964fad1/plone/scale/storage.py#L69

AFAIR PersistentDict are completely rewritten when something changes.

I guess that using OOBTree directly is a faster option (which comes with builtin conflict resolution that does not require you add the plone.scale to the zeo script).

**Comment by ale-rt** [comment]
See also the discussion happening here https://github.com/plone/plone.scale/issues/41

**Comment by mamico** [comment]
> Just my 2 c.
> 
> I think ScalesDict is not performant because it inherits from PersistentDict...

@alert, thank you for your suggestion. But I assume it’s another story: here the main performance improvement is not about storing/reading data, but the removal of `IDisableCSRFProtection` (which should also be a possible security improvement).

**Comment by ale-rt** [comment]
> > Just my 2 c.
> > I think ScalesDict is not performant because it inherits from PersistentDict...
> 
> @Alert, thank you for your suggestion. But I assume it’s another story: here the main performance improvement is not about storing/reading data, but the removal of `IDisableCSRFProtection` (which should also be a possible security improvement).

Yeah, I see your point, but I think that if performance is what you are looking for switching away from `PersistentDict` is definitely something worth considering.

Anyway, it looks good but of course this is not merge-able with that patch.

**Comment by mamico** [comment]
> Anyway, it looks good but of course this is not merge-able with that patch.

Absolutely, it's what I've tried to  say before. If the main idea is acceptable to all reviewers, I'll open a second PR in `plone.scale`. If you prefer, I can also remove the other optimizations except the `IDisableCSRFProtection` removing.

**Comment by ale-rt** [comment]
> > Anyway, it looks good but of course this is not merge-able with that patch.
> 
> Absolutely, it's what I've tried to say before. If the main idea is acceptable to all reviewers, I'll open a second PR in `plone.scale`. If you prefer, I can also remove the other optimizations except the `IDisableCSRFProtection` removing.

Of course it is accessible

**Comment by mauritsvanrees** [comment]
I have left some comments on the commit with the temporary monkey patch (f71706a2f9b405b4902b658a75837c405abd5401).

The approach seems good. Please continue. :-)

**Comment by mauritsvanrees** [comment]
Thanks, sounds logical now.
I have merged your `plone.scale` PR.

**Comment by mamico** [comment]
@jenkins-plone-org please run jobs

**Comment by mauritsvanrees** [comment]
One known failing flaky robot test in one Jenkins job. So should be fine to merge.

@ale-rt Can you do a final review?

**Review by ale-rt — APPROVED** [review]
LGTM!

**Comment by mauritsvanrees** [comment]
I have merged this, but GitHub UI does not show it, seems a bit slow the last few minutes.

Anyway, thanks for your work @mamico and thanks for the review @ale-rt !

**Pull request closed without merging**
